src/02-common-principles.md

```diff
@@ -167,15 +162,120 @@ my_dataset/
+source, and derived data while maintaining BIDS compliancy of the raw data
+folder. When using this convention it is RECOMMENDED to set the `SourceDatasets`
+field in `dataset_description.json` of each subfolder of `derivatives` to:
+
+```JSON
+{
+  "SourceDatasets": [ {"URL": "file://../../rawdata/"} ]
+}
+```
+
+### Storage of derived datasets
+
+Derivatives can be stored/distributed in two ways:
+
+1.  Under a `derivatives/` subfolder in the root of the source BIDS dataset
+    folder to make a clear distinction between raw data and results of data
+    processing. A data processing pipeline will typically have a dedicated directory
+    under which it stores all of its outputs. Different components of a pipeline can,
+    however, also be stored under different subfolders. There are few restrictions on
+    the directory names; it is RECOMMENDED to use the format `<pipeline>-<variant>` in
+    cases where it is anticipated that the same pipeline will output more than one variant (e.g.,
+    `AFNI-blurring`, `AFNI-noblurring`, etc.). For the sake of consistency, the
+    subfolder name SHOULD be the `GeneratedBy.Name` field in
+    `data_description.json`, optionally followed by a hyphen and a suffix (see
+    [Derived dataset and pipeline description][derived-dataset-description].
+
+    Example of derivatives with one directory per pipeline:
+
+    ```Plain
+    <dataset>/derivatives/fmriprep-v1.4.1/sub-0001
+    <dataset>/derivatives/spm/sub-0001
+    <dataset>/derivatives/vbm/sub-0001
+    ```
+
+    Example of a pipeline with split derivative directories:
+
+    ```Plain
+    <dataset>/derivatives/spm-preproc/sub-0001
+    <dataset>/derivatives/spm-stats/sub-0001
+    ```
+
+    Example of a pipeline with nested derivative directories:
+
+    ```Plain
+    <dataset>/derivatives/spm-preproc/sub-0001
+    <dataset>/derivatives/spm-preproc/derivatives/spm-stats/sub-0001
+    ```
+
+
+1.  As a standalone dataset independent of the source (raw or derived) BIDS
+    dataset.
+    This way of specifying derivatives is particularly useful when the source
+    dataset is provided with read-only access, for publishing derivatives as
+    independent bodies of work, or for describing derivatives that were created
+    from more than one source dataset.
+    The `sourcedata/` subdirectory MAY be used to include the source dataset(s)
+    that were used to generate the derivatives.
+    Likewise, any code used to generate the derivatives from the source data
+    MAY be included in the `code/` subdirectory.
+
+    Example of a derivative dataset including the raw dataset as source:
+
+    ```Plain
+    my_processed_data/
+      code/
+        processing_pipeline-1.0.0.img
+        hpc_submitter.sh
+        ...
+      sourcedata/
+        dataset_description.json
+        participants.tsv
+        sub-01/
+        sub-02/
+        ...
+      dataset_description.json
+      sub-01/
+      sub-02/
+      ...
+    ```
```

Comment:
Copying @Lestropie's comment from https://github.com/bids-standard/bids-specification/pull/488#pullrequestreview-422282063:

> Given the differential definitions of "source" vs. "raw" datasets in "common principles", it might be better to avoid use of the word "source" in the context of "data from which the derivatives were generated". That or "source" in reference to pre-BIDS-raw data would need to be changed.

I would be okay with an alternate name, but the `sourcedata/` keyword is already there, means the data that were transformed to produce the current dataset, and constitutes a boundary between datasets.

> There's also some potential ambiguity here in terms of hierarchical directory structure. It's permissible for derivative data to be stored in a sub-directory of the raw dataset, but it's also possible for the raw data to be stored in a sub-directory of the derivative dataset, and none of the directory names are prescribed; so it's a bit of a free-for-all, with the only restriction being storage of raw and derived data at the same hierarchical level.

Yes. I expect you'll end up with chains of derivatives or chains of sources, but not generally both within the same dataset. Extrapolating purely from my own preferences, I would suggest that the derivatives-as-subdirectory is a pre-derivatives hack that will largely go away in the future. Keeping a versioned record of the precise inputs of a derivative is a much more useful way to link datasets. (See: https://doi.org/10.7490/f1000research.1116363.1)

> This would mean that for automatic parsing, unambiguously determining what's raw and what's derived would require running the entire validator on each candidate directory.

I had been thinking that we might need something in the `dataset_description.json` to unambiguously declare a dataset either derivative or raw. At the moment, `GeneratedBy` is a distinguishing feature, but if #369 is accepted, that won't work. I will make a comment to the main thread.

> Also, if the raw dataset is provided read-only, writing a read-write version of it along with the derivative data could be seen to be counteracting the intent of that read-only provision.

I think in practice, these are unenforceable, but if they were, the subdirectory can be `chmod`ded `-w`.

> If the raw data are to be in some way distributed alongside the derivative data, doing so according to the suggestion provided in "common principles":
> 
> ```
> my_dataset/
>   sourcedata/
>     ...
>   rawdata/
>     dataset_description.json
>     participants.tsv
>     sub-01/
>     sub-02/
>     ...
>   derivatives/
>     ...
> ```
> 
> may be more faithful to such.
> 
> Actually looking at it again, this is in (**edit**: the old version of) "Common principles":
> 
> > These data MUST be kept in separate sourcedata and derivatives folders
> 
> Permitting BIDS-raw data to be stored as a sub-directory within a BIDS-derivatives directory could arguably be seen as contradictory to this requirement, since the derivatives directory is no longer "separate" to the (duplicate of the) raw data but a superset thereof.
> 
> So personally I'd probably err on _not_ permitting raw data as sub-directories of derived data.

`derivatives/` and `sourcedata/` have to constitute boundaries between datasets, or you really can't coherently parse a dataset. What happens on the other side of those boundaries is out of our control. I have seen in practice people nesting input datasets inside output datasets, so this is just trying to give guidance on a consistent, BIDS-y way to do that.

Reply:
No particular contentions with any of the above from me.

Only other possibility to float in this context would be whether we expect *a priori* / have learned from experience that a particular inter-dataset filesystem arrangement is likely to lead to fewer problems / less ambiguity, and would consequently warrant being the recommended approach, but without restriction on the alternatives. If anything this might actually make things easier for beginners, as they would no longer be obliged to naively make such a decision up-front.

Reply:
I'm not sure there currently exists a consensus on the best way to do it. I obviously have my opinion, but I'm not sure that we should really recommend one, as I think it's going to depend on the use case. For example, some approaches may be more conducive to doing processing while others are better for packaging up and distributing.

Hopefully this makes clear that datasets should be independently distributable. If that's the case, then their relative locations should be irrelevant, and moving to an alternative (or no) nesting should be more-or-less trivial.

Happy to consider any wording you might suggest that will get these ideas across. I think I'm running out of generative capacity at the moment.